packages/collaboration-manager/src/CollaborationManager.ts

```diff
@@ -212,34 +226,63 @@ export class CollaborationManager {
       return;
     }
 
+    /**
+     * If operation is local, send it to the server
+     */
     if (operation.userId === this.#config.userId) {
       void this.#client?.send(operation);
     } else {
+      /**
+       * If operation is remote, transform undo/redo stacks
+       */
+      this.#undoRedoManager.transformStacks(operation);
+
+      /**
+       * If we got a new remote operation - transform current batch
+       * If batch is empty leave it as it is
+       */
+      this.#currentBatch = this.#currentBatch?.transform(operation) ?? null;
+
       return;
     }
 
-    const onBatchTermination = (batch: OperationsBatch, lastOp?: Operation): void => {
-      const effectiveOp = batch.getEffectiveOperation();
+    if (!this.#shouldHandleEvents) {
+      return;
+    }
 
-      if (effectiveOp) {
-        this.#undoRedoManager.put(effectiveOp);
-      }
+    /**
+     * If there is no current batch, create a new one with current operation
+     */
+    if (this.#currentBatch === null) {
+      this.#currentBatch = new BatchedOperation(operation);
 
-      /**
-       * lastOp is the operation on which the batch was terminated.
-       * So if there is one, we need to create a new batch
-       *
-       * lastOp could be null if the batch was terminated by time out
-       */
-      this.#currentBatch = lastOp === undefined ? null : new OperationsBatch(onBatchTermination, lastOp);
-    };
+      return;
+    }
 
-    if (this.#currentBatch === null) {
-      this.#currentBatch = new OperationsBatch(onBatchTermination, operation);
+    /**
+     * If current operation could not be added to the batch, then terminate current batch and create a new one with current operation
+     *
+     * @todo - add debounce timeout 500ms
+     */
+    if (!this.#currentBatch.canAdd(operation)) {
+      this.#undoRedoManager.put(this.#currentBatch);
+
+      this.#currentBatch = new BatchedOperation(operation);
 
       return;
     }
 
     this.#currentBatch.add(operation);
   }
+
+  /**
+   * Puts current batch to the undo stack and clears the batch
+   */
+  #emptyBatch(): void {
```

Comment:
I don't really like the name as it implies the current batch is being emptied. Also it doesn't show that the current batch is being added to the undo redo manager